arch/risc-v/src/esp32c3/esp32c3_spiflash.c

```diff
@@ -0,0 +1,745 @@
+
+/* MTD driver methods */
+
+static int esp32c3_erase(struct mtd_dev_s *dev, off_t startblock,
+                         size_t nblocks);
+static ssize_t esp32c3_read(struct mtd_dev_s *dev, off_t offset,
+                            size_t nbytes, uint8_t *buffer);
+static ssize_t esp32c3_read_decrypt(struct mtd_dev_s *dev,
+                                    off_t offset,
+                                    size_t nbytes,
+                                    uint8_t *buffer);
+static ssize_t esp32c3_bread(struct mtd_dev_s *dev, off_t startblock,
+                             size_t nblocks, uint8_t *buffer);
+static ssize_t esp32c3_bread_decrypt(struct mtd_dev_s *dev,
+                                     off_t startblock,
+                                     size_t nblocks,
+                                     uint8_t *buffer);
+static ssize_t esp32c3_write(struct mtd_dev_s *dev, off_t offset,
+                             size_t nbytes, const uint8_t *buffer);
+static ssize_t esp32c3_bwrite(struct mtd_dev_s *dev, off_t startblock,
+                              size_t nblocks, const uint8_t *buffer);
+static ssize_t esp32c3_bwrite_encrypt(struct mtd_dev_s *dev,
+                                      off_t startblock,
+                                      size_t nblocks,
+                                      const uint8_t *buffer);
+static int esp32c3_ioctl(struct mtd_dev_s *dev, int cmd,
+                         unsigned long arg);
+
+/****************************************************************************
+ * Private Data
+ ****************************************************************************/
+
+static struct esp32c3_spiflash_s g_esp32c3_spiflash =
+{
+  .mtd =
+          {
+            .erase  = esp32c3_erase,
+            .bread  = esp32c3_bread,
+            .bwrite = esp32c3_bwrite,
+            .read   = esp32c3_read,
+            .ioctl  = esp32c3_ioctl,
+#ifdef CONFIG_MTD_BYTE_WRITE
+            .write  = esp32c3_write,
+#endif
+            .name   = "esp32c3_spiflash"
+          },
+  .chip = &g_rom_flashchip,
+};
+
+static struct esp32c3_spiflash_s g_esp32c3_spiflash_encrypt =
+{
+  .mtd =
+          {
+            .erase  = esp32c3_erase,
+            .bread  = esp32c3_bread_decrypt,
+            .bwrite = esp32c3_bwrite_encrypt,
+            .read   = esp32c3_read_decrypt,
+            .ioctl  = esp32c3_ioctl,
+#ifdef CONFIG_MTD_BYTE_WRITE
+            .write  = NULL,
+#endif
+            .name   = "esp32c3_spiflash_encrypt"
+          }
+};
+
+/* Ensure exclusive access to the driver */
+
+static sem_t g_exclsem = SEM_INITIALIZER(1);
+
+/****************************************************************************
+ * Private Functions
+ ****************************************************************************/
+
+/****************************************************************************
+ * Name: esp32c3_erase
+ *
+ * Description:
+ *   Erase SPI Flash designated sectors.
```

Comment:
This is a bit confusing. Usually sectors and blocks do not relate to the same thing. What does this function actually receive as input parameters?

Reply:
Sectors, Blocks and Offsets are mixed here: basically a block refer to the index of a sector I want to erase and nblocks the amount of sectors I want to erase. The offset is the physical position of the sector, just multiplying startblock * ERASE_SIZE I got the offset.